zipkin-storage/cassandra-v1/src/main/java/zipkin2/storage/cassandra/v1/TraceIdIndexer.java

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright 2015-2020 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin2.storage.cassandra.v1;
+
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.DelayQueue;
+import java.util.concurrent.Delayed;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import zipkin2.internal.Nullable;
+import zipkin2.storage.cassandra.v1.IndexTraceId.Input;
+
+/**
+ * This ensures we only attempt to write rows that would extend the timestamp range of a trace index
+ * query, by de-duplicating redundant entries upon {@link #iterator()}
+ *
+ * <p>This is shared singleton as inserts can come from any thread.
+ */
+interface TraceIdIndexer extends Iterable<Input> {
+  TraceIdIndexer NOOP = new TraceIdIndexer() {
+    @Override public void add(Input input) {
+    }
+
+    @Override public Iterator<Input> iterator() {
+      return Collections.emptyIterator();
+    }
+  };
+
+  void add(Input input);
+
+  /** This is shared singleton as inserts can come from any thread. */
+  class Factory {
+    final ConcurrentMap<Entry<String, Long>, Expiration<Entry<String, Long>, Pair>> cache =
+      new ConcurrentHashMap<>();
+    final DelayQueue<Expiration<Entry<String, Long>, Pair>> expirations = new DelayQueue<>();
+    final long ttlNanos;
+    final int cardinality;
+    final String table;
+
+    Factory(String table, long ttlNanos, int cardinality) {
+      this.table = table;
+      this.ttlNanos = ttlNanos;
+      this.cardinality = cardinality;
+    }
+
+    long nanoTime() {
+      return System.nanoTime();
+    }
+
+    <K, V> Expiration<K, V> newExpiration(K key, V value) {
+      return new Expiration<>(this, key, value, nanoTime() + ttlNanos);
+    }
+
+    TraceIdIndexer newIndexer() {
+      trimCache(); // We let the cache overfill per call to SpanConsumer.accept. This trims it back
```

Comment:
So IIUC, we clean up anytime we index a new batch of spans. May as well add a line of comment with that since it's not so obvious

Reply:
sure